groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -702,58 +1005,373 @@ struct Optional_Data<TYPE, true> : public Optional_DataImp<TYPE> {
+        // Use the currently installed default allocator to supply memory.
+        // 'original' is left in a valid but unspecified state.
+
+    template <class ANY_TYPE>
+    explicit optional(
+        optional<ANY_TYPE>&& original,
+        BSLSTL_OPTIONAL_DECLARE_IF_NOT_SAME(TYPE, ANY_TYPE),
+        BSLSTL_OPTIONAL_DECLARE_IF_CONSTRUCTS_FROM_BSL_OPTIONAL(TYPE,
+                                                                ANY_TYPE),
+        BSLSTL_OPTIONAL_DECLARE_IF_CONSTRUCT_DOES_NOT_PROPAGATE_ALLOCATOR(
+            TYPE,
+            ANY_TYPE),
+        BSLSTL_OPTIONAL_DECLARE_IF_EXPLICIT_CONSTRUCT(TYPE, ANY_TYPE));
+        // Create a disengaged 'optional' object if the specified 'original'
+        // object is disengaged, and an 'optional' object with the value of
+        // 'original.value()' (of 'ANY_TYPE') converted to 'TYPE' otherwise.
+        // Use the currently installed default allocator to supply memory.
+        // 'original' is left in a valid but unspecified state.
+
+    template <class ANY_TYPE>
+    optional(
+        optional<ANY_TYPE>&& original,
+        BSLSTL_OPTIONAL_DECLARE_IF_NOT_SAME(TYPE, ANY_TYPE),
+        BSLSTL_OPTIONAL_DECLARE_IF_CONSTRUCTS_FROM_BSL_OPTIONAL(TYPE,
+                                                                ANY_TYPE),
+        BSLSTL_OPTIONAL_DECLARE_IF_CONSTRUCT_PROPAGATES_ALLOCATOR(TYPE,
+                                                                  ANY_TYPE),
+        BSLSTL_OPTIONAL_DECLARE_IF_NOT_EXPLICIT_CONSTRUCT(TYPE, ANY_TYPE));
+        // Create a disengaged 'optional' object if the specified 'original'
+        // object is disengaged, and an 'optional' object with the value of
+        // 'original.value()' otherwise.  This is a special case constructor
+        // where 'ANY_TYPE' is a non-const version of 'TYPE' and we use the
+        // allocator from 'original' to supply memory.  'original' is left in a
+        // valid but unspecified state.
+
+    template <class ANY_TYPE>
+    explicit optional(
+        optional<ANY_TYPE>&& original,
+        BSLSTL_OPTIONAL_DECLARE_IF_NOT_SAME(TYPE, ANY_TYPE),
+        BSLSTL_OPTIONAL_DECLARE_IF_CONSTRUCTS_FROM_BSL_OPTIONAL(TYPE,
+                                                                ANY_TYPE),
+        BSLSTL_OPTIONAL_DECLARE_IF_CONSTRUCT_PROPAGATES_ALLOCATOR(TYPE,
+                                                                  ANY_TYPE),
+        BSLSTL_OPTIONAL_DECLARE_IF_EXPLICIT_CONSTRUCT(TYPE, ANY_TYPE));
+        // Create a disengaged 'optional' object if the specified 'original'
+        // object is disengaged, and an 'optional' object with the value of
+        // 'original.value()' otherwise.  This is a special case constructor
+        // where 'ANY_TYPE' is a non-const version of 'TYPE' and we use the
+        // allocator from 'original' to supply memory.  'original' is left in a
+        // valid but unspecified state.
+
+#else
+    template <class ANY_TYPE>
```

Comment:
If deduction cannot happen, then can these constructors ever be called?  If not - is there any reason for them existing at all?

[I'll stop here and will continue Monday]

Reply:
I don't understand this comment. What do you mean by deduction can't happen ?

Reply:
Well, this code won't compile (https://bbgb.dev.bloomberg.com/z/Mm_5oy)

```
#include <bslmf_movableref.h>

using namespace BloombergLP;

struct S {
    template<class T>
    S(bslmf::MovableRef<T> t)
    {}
};


void g()
{
    int x;
    S s(bslmf::MovableRefUtil::move(x));
}
```

Reply:
it doesn't compile in C++11 and other newer modes because the deduction problem is with the MovableRef and C++11, it compiles fine in C++03. Note that deduction works just fine if you were to use an rvalue ref in the S's template constructor instead of MovableRef. This is exactly why we have two sets of overloads - one using a movableRef for C++03, and one using a naked rvalue reference for C++11. There is a comment explaining this slightly above (line 1190 for me), I'm going to guess it isn't very clear. :) Feel free to suggest a better explanation.

Reply:
I had a feeling my brain was entering Friday evening shutdown... I remembered that MovableRef doesn't allow type deduction, but forgot it only does this in C++11+.  I'll go back to this with a fresh head on Monday.  My apologies.